### PR TITLE
fix: only package on deploy if necessary

### PR DESCRIPTION
**Pull request by TheSriram**
- if supplied template, has any resource, metadata, globals that already
  have a s3 url, do not repackage.

*Issue #, if available:*

*Description of changes:*

*Checklist:*

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/lib/package/artifact_exporter.py [hunk 1]
```diff
@@ -263,6 +263,17 @@ def export(self, resource_id, resource_dict, parent_dir):
...
+    def get_property(self, resource_dict):
+        return jmespath.search(self.PROPERTY_NAME, resource_dict)
+
+    def is_pre_exported(self, template_dict):
+        property_value = self.get_property(template_dict if template_dict else {})
+
+        if not property_value and not self.PACKAGE_NULL_PROPERTY:
+            return False
+
+        return is_s3_url(property_value)
```
Comment (sanathkr) [inline samcli/lib/package/artifact_exporter.py]: what if self.PACKAGE_NULL_PROPERTY is set?

File: samcli/commands/deploy/command.py [hunk 2]
```diff
@@ -193,10 +193,12 @@ def do_cli(
...
             profile=profile,
         ) as package_context:
-            package_context.run()
+            pre_packaged = package_context.is_pre_packaged(template_path=template_file)
+            if not pre_packaged:
+                package_context.run()
 
         with DeployContext(
-            template_file=output_template_file.name,
+            template_file=template_file if packaged else output_template_file.name,
```
[the remark this answers is not in the document]
Comment (TheSriram) [reply samcli/commands/deploy/command.py]: ah, I missed this when I changed variable to pre_packaged.

File: samcli/lib/package/artifact_exporter.py (hunk 1 above)
Comment (TheSriram) [reply samcli/lib/package/artifact_exporter.py]: we pass a None to is_s3_url which leads to a valuerror, rendering it false.

File: samcli/lib/package/artifact_exporter.py [hunk 3]
```diff
@@ -621,3 +640,90 @@ def export(self):
...
+                if exporter_class.RESOURCE_TYPE != resource_type:
+                    continue
+
+                # Check for pre-exported code resources
+                exporter = exporter_class(self.uploader)
+                pre_exported_resources.append(exporter.is_pre_exported(template_dict=resource_dict))
+
+        return any(pre_exported_resources)
+
+    def contains_packaged_artifacts(self):
```
Comment (TheSriram) [inline samcli/lib/package/artifact_exporter.py]: This is the main entry point to check for packaged artifacts.

**Comment by TheSriram** [comment]
rebased onto latest from v0.32.0 branch

**Comment by TheSriram** [comment]
Closing this PR.

**Comment by wmonk** [comment]
Did this fix ever land? It's super frustrating making config changes while using an `s3` url for `CodeUri` and it repackaging each time.

**Pull request closed without merging**